Mathlib/Topology/Algebra/Order/LiminfLimsup.lean

```diff
@@ -300,54 +300,124 @@ end CompleteLinearOrder
 
 end LiminfLimsup
 
-section Monotone
-
-variable {ι R S : Type _} {F : Filter ι} [NeBot F] [CompleteLinearOrder R] [TopologicalSpace R]
-  [OrderTopology R] [CompleteLinearOrder S] [TopologicalSpace S] [OrderTopology S]
-
-/-- An antitone function between complete linear ordered spaces sends a `Filter.limsSup`
-to the `Filter.liminf` of the image if it is continuous at the `limsSup`. -/
-theorem Antitone.map_limsSup_of_continuousAt {F : Filter R} [NeBot F] {f : R → S}
+section Monotone'
+
+-- Q: Should the following lemmas be avoided altogether? If not, what is the right place for them?
+
+lemma Monotone.isCoboundedUnder_ge [Preorder X] [Preorder Y] {f : X → Y} (hf : Monotone f)
+    {u : ι → X} (F : Filter ι) [NeBot F] (bdd : F.IsBoundedUnder (· ≤ ·) u) :
+    F.IsCoboundedUnder (· ≥ ·) (f ∘ u) := by
+  apply Filter.IsBounded.isCobounded_flip
+  obtain ⟨b, hb⟩ := bdd
+  refine ⟨f b, ?_⟩
+  simp only [eventually_map, Function.comp_apply] at hb ⊢
+  filter_upwards [hb] with i hi using hf hi
+
+lemma Monotone.isCoboundedUnder_le [Preorder X] [Preorder Y] {f : X → Y} (hf : Monotone f)
+    {u : ι → X} (F : Filter ι) [NeBot F] (bdd : F.IsBoundedUnder (· ≥ ·) u) :
+    F.IsCoboundedUnder (· ≤ ·) (f ∘ u) :=
+  @Monotone.isCoboundedUnder_ge Xᵒᵈ Yᵒᵈ ι _ _ f hf.dual u F _ bdd
+
+lemma Antitone.isCoboundedUnder_le [Preorder X] [Preorder Y] {f : X → Y} (hf : Antitone f)
+    {u : ι → X} (F : Filter ι) [NeBot F] (bdd : F.IsBoundedUnder (· ≤ ·) u) :
+    F.IsCoboundedUnder (· ≤ ·) (f ∘ u) :=
+  @Monotone.isCoboundedUnder_ge X Yᵒᵈ ι _ _ f hf u F _ bdd
+
+lemma Antitone.isCoboundedUnder_ge [Preorder X] [Preorder Y] {f : X → Y} (hf : Antitone f)
+    {u : ι → X} (F : Filter ι) [NeBot F] (bdd : F.IsBoundedUnder (· ≥ ·) u) :
+    F.IsCoboundedUnder (· ≥ ·) (f ∘ u) :=
+  @Monotone.isCoboundedUnder_le X Yᵒᵈ ι _ _ f hf u F _ bdd
+
+variable {ι R S : Type _} {F : Filter ι} [NeBot F]
+  [ConditionallyCompleteLinearOrder R] [TopologicalSpace R] [OrderTopology R]
+  [ConditionallyCompleteLinearOrder S] [TopologicalSpace S] [OrderTopology S]
+
+-- Q: Was there a way to "automatically specialize" these kinds of lemmas from
+-- conditionally complete linear orders to complete linear orders so that the then
+-- trivial boundedness assumptions (such as `F.IsBounded (· ≤ ·)`, `F.IsBounded (· ≥ ·)`)
+-- don't need to be provided by the user?
+--
+-- That would allow to use the new `Antitone.map_limsSup_of_continuousAt'` (with prime) and friends
+-- as strictly more general direct replacements of `Antitone.map_limsSup_of_continuousAt` (no
+-- prime) and friends.
```

Comment:
I think you can use something like `(h : f.IsBoundedUnder (· ≤ ·) u := by isBoundedDefault)`. See [tendsto_of_liminf_eq_limsup](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Order/LiminfLimsup.html#tendsto_of_liminf_eq_limsup) for example.

Reply:
If it works indeed you should replace the existing versions.

Reply:
Thank you! This seems to work. I'm renaming the new ones to see if they now work throughout Mathlib as direct replacements of the old ones.

Reply:
If I get to removing the old ones altogether, is it ok to keep the `#align`s although the type signature will have changed? Since the results would presumably still work with the same _explicit_ arguments, my naive opinion is that aligning would still be ok.

Reply:
Yes that's fine, because 1) indeed this is close enough in practice and 2) we got past `port-complete` so we don't need to be as careful as before IIUC.

Reply:
The tactic `isBoundedDefault` was for filters only (as far as I understood). I wrote a similar 2-line macro `setIsBddDefault` to be used with default boundedness of sets.

Reply:
Oh, or would it be desirable that the same tactic handles boundedness of both filters and sets in complete linear orders? The tactic `isBoundedDefault` seemed to be in `Filter` namespace, so I thought boundedness of sets would not be in scope, but now I'm not so sure anymore...

Reply:
Yes I think the best solution is to have the same tactic taking care of both. That said, I've thought about it a few times before, and arrived to the conclusion that there is a "good" reason that we don't have the set version: even if we had, we wouldn't be able to merge the APIs of `CompleteLattice` and `ConditionallyCompleteLattice`, because we still can't get rid of the `Nonempty` assumptions that are required in the conditionally complete case but not in the complete case. This situation doesn't happen for (nontrivial?) filters, so here it makes more sense to have each lemma once.

That doesn't mean that we shouldn't add it, in fact it can't hurt and I have some ideas about a (gigantic) refactor that would make it useful, I just wanted to mention why it's not so strange that it is missing.

Reply:
On the technical side, I'm not sure how to make it so that the same tactic work without messing up with the import tree. We could probably take inspiration about `positivity` that is split into small extensions, but if that gets too hard then having two tactics is not that bad (you basically never have to call these tactics explicitly anyway)

Reply:
I think having two tactics is fine for now. If this kind of thing keeps coming up we can develop a uniform way to deal with it.